File: tracker/progress.py
from __future__ import annotations

import json
import sqlite3
from contextlib import contextmanager
from datetime import date, datetime, timedelta
from typing import Any, Dict, Iterable, List, Optional

from tracker.paths import progress_db_path

SUBJECTS = ("dsa", "lld", "hld", "ai", "reading", "walk", "review", "other")


@contextmanager
def connect():
    path = progress_db_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    conn = sqlite3.connect(str(path))
    conn.row_factory = sqlite3.Row
    try:
        init_db(conn)
        yield conn
        conn.commit()
    finally:
        conn.close()


def init_db(conn: sqlite3.Connection) -> None:
    conn.executescript(
        """
        CREATE TABLE IF NOT EXISTS sessions (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            ts TEXT NOT NULL,
            subject TEXT NOT NULL,
            minutes INTEGER NOT NULL DEFAULT 0,
            notes TEXT NOT NULL DEFAULT '',
            problems_count INTEGER NOT NULL DEFAULT 0,
            extra TEXT NOT NULL DEFAULT '{}'
        );

        CREATE TABLE IF NOT EXISTS weekly_reviews (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            week_start TEXT NOT NULL UNIQUE,
            created_at TEXT NOT NULL,
            dsa TEXT NOT NULL DEFAULT '',
            lld TEXT NOT NULL DEFAULT '',
            hld TEXT NOT NULL DEFAULT '',
            ai TEXT NOT NULL DEFAULT '',
            personal TEXT NOT NULL DEFAULT ''
        );

        CREATE TABLE IF NOT EXISTS settings (
            key TEXT PRIMARY KEY,
            value TEXT NOT NULL
        );
        """
    )


def week_start_iso(moment: datetime) -> str:
    monday = moment.date() - timedelta(days=moment.weekday())
    return monday.isoformat()


def add_session(
    subject: str,
    minutes: int = 0,
    notes: str = "",
    problems_count: int = 0,
    extra: Optional[Dict[str, Any]] = None,
    ts: Optional[str] = None,
) -> Dict[str, Any]:
    subject = (subject or "other").lower()
    if subject not in SUBJECTS:
        subject = "other"
    payload = {
        "ts": ts or datetime.now().isoformat(timespec="seconds"),
        "subject": subject,
        "minutes": max(0, int(minutes or 0)),
        "notes": notes or "",
        "problems_count": max(0, int(problems_count or 0)),
        "extra": json.dumps(extra or {}),
    }
    with connect() as conn:
        cursor = conn.execute(
            """
            INSERT INTO sessions (ts, subject, minutes, notes, problems_count, extra)
            VALUES (:ts, :subject, :minutes, :notes, :problems_count, :extra)
            """,
            payload,
        )
        payload["id"] = cursor.lastrowid
    return payload


def save_review(week_start: str, body: Dict[str, str]) -> Dict[str, Any]:
    record = {
        "week_start": week_start,
        "created_at": datetime.now().isoformat(timespec="seconds"),
        "dsa": body.get("dsa") or "",
        "lld": body.get("lld") or "",
        "hld": body.get("hld") or "",
        "ai": body.get("ai") or "",
        "personal": body.get("personal") or "",
    }
    with connect() as conn:
        conn.execute(
            """
            INSERT INTO weekly_reviews (week_start, created_at, dsa, lld, hld, ai, personal)
            VALUES (:week_start, :created_at, :dsa, :lld, :hld, :ai, :personal)
            ON CONFLICT(week_start) DO UPDATE SET
                created_at=excluded.created_at,
                dsa=excluded.dsa,
                lld=excluded.lld,
                hld=excluded.hld,
                ai=excluded.ai,
                personal=excluded.personal
            """,
            record,
        )
    return record


def recent_sessions(limit: int = 20) -> List[Dict[str, Any]]:
    with connect() as conn:
        rows = conn.execute(
            "SELECT * FROM sessions ORDER BY ts DESC, id DESC LIMIT ?",
            (limit,),
        ).fetchall()
    return [dict(row) for row in rows]


def get_review(week_start: str) -> Optional[Dict[str, Any]]:
    with connect() as conn:
        row = conn.execute(
            "SELECT * FROM weekly_reviews WHERE week_start = ?",
            (week_start,),
        ).fetchone()
    return dict(row) if row else None


def stats_for_week(moment: datetime) -> Dict[str, Any]:
    start = week_start_iso(moment)
    end_date = date.fromisoformat(start) + timedelta(days=7)
    with connect() as conn:
        rows = conn.execute(
            "SELECT * FROM sessions WHERE ts >= ? AND ts < ?",
            (start, end_date.isoformat()),
        ).fetchall()
        total_row = conn.execute(
            "SELECT COALESCE(SUM(problems_count), 0) AS n FROM sessions WHERE subject = 'dsa'"
        ).fetchone()
    by_subject: Dict[str, Dict[str, int]] = {}
    walk_days = set()
    reading_days = set()
    for row in rows:
        subject = row["subject"]
        bucket = by_subject.setdefault(subject, {"minutes": 0, "sessions": 0, "problems": 0})
        bucket["minutes"] += int(row["minutes"] or 0)
        bucket["sessions"] += 1
        bucket["problems"] += int(row["problems_count"] or 0)
        day = str(row["ts"])[:10]
        if subject == "walk":
            walk_days.add(day)
        if subject == "reading":
            reading_days.add(day)
    return {
        "week_start": start,
        "by_subject": by_subject,
        "walk_days": len(walk_days),
        "reading_days": len(reading_days),
        "dsa_problems_total": int(total_row["n"] if total_row else 0),
        "dsa_problems_week": by_subject.get("dsa", {}).get("problems", 0),
        "study_minutes_week": sum(
            by_subject.get(key, {}).get("minutes", 0) for key in ("dsa", "lld", "hld", "ai")
        ),
        "review": get_review(start),
    }


def set_setting(key: str, value: str) -> None:
    with connect() as conn:
        conn.execute(
            "INSERT INTO settings(key, value) VALUES(?, ?) ON CONFLICT(key) DO UPDATE SET value=excluded.value",
            (key, value),
        )


def get_setting(key: str, default: str = "") -> str:
    with connect() as conn:
        row = conn.execute("SELECT value FROM settings WHERE key = ?", (key,)).fetchone()
    return str(row["value"]) if row else default

File: src/lib/progress.ts
import { and, desc, eq, gte, lt } from "drizzle-orm";
import { db } from "@/db";
import { sessions, weeklyReviews } from "@/db/schema";
import { loadConfig } from "./config";
import { weekDsaProblemStats } from "./dsa-problems";
import {
  addCalendarDays,
  pythonWeekday,
  ymdInZone,
  zonedParts,
  zonedWallToDate,
} from "./timezone";
import { listProblems } from "./zettel/notes";
import type {
  ReviewRecord,
  SessionRecord,
  Subject,
  SubjectBucket,
  WeekStats,
} from "./types";
import { SUBJECTS } from "./types";

function isSubject(value: string): value is Subject {
  return (SUBJECTS as readonly string[]).includes(value);
}

function serializeSession(row: typeof sessions.$inferSelect): SessionRecord {
  return {
    id: row.id,
    ts: row.ts.toISOString(),
    subject: row.subject,
    minutes: row.minutes,
    notes: row.notes,
    problems_count: row.problemsCount,
    extra: row.extra ?? {},
  };
}

function serializeReview(row: typeof weeklyReviews.$inferSelect): ReviewRecord {
  return {
    id: row.id,
    week_start: row.weekStart,
    created_at: row.createdAt.toISOString(),
    dsa: row.dsa,
    lld: row.lld,
    hld: row.hld,
    ai: row.ai,
    personal: row.personal,
  };
}

export function weekStartIso(moment: Date, timeZone = loadConfig().timezone): string {
  const local = zonedParts(moment, timeZone);
  const weekday = pythonWeekday(local.year, local.month, local.day);
  const monday = addCalendarDays(local.year, local.month, local.day, -weekday);
  return `${monday.year}-${String(monday.month).padStart(2, "0")}-${String(monday.day).padStart(2, "0")}`;
}

export async function addSession(input: {
  subject: string;
  minutes?: number;
  notes?: string;
  problems_count?: number;
  extra?: Record<string, unknown>;
  ts?: string;
}): Promise<SessionRecord> {
  let subject = (input.subject || "other").toLowerCase();
  if (!isSubject(subject)) subject = "other";

  const ts = input.ts ? new Date(input.ts) : new Date();
    const [row] = await db
      .insert(sessions)
      .values({
        ts,
        subject,
        minutes: Math.max(0, Math.trunc(input.minutes || 0)),
        notes: input.notes || "",
        problemsCount: Math.max(0, Math.trunc(input.problems_count || 0)),
        extra: input.extra || {},
      })
      .returning();

    if (!row) throw new Error("Failed to insert session");
    return serializeSession(row);
}

export async function deleteSession(id: string): Promise<SessionRecord | null> {
  const [row] = await db.delete(sessions).where(eq(sessions.id, id)).returning();
  return row ? serializeSession(row) : null;
}

export async function saveReview(
  weekStart: string,
  body: {
    dsa?: string;
    lld?: string;
    hld?: string;
    ai?: string;
    personal?: string;
  },
): Promise<ReviewRecord> {
  const createdAt = new Date();
  const values = {
    weekStart,
    createdAt,
    dsa: body.dsa || "",
    lld: body.lld || "",
    hld: body.hld || "",
    ai: body.ai || "",
    personal: body.personal || "",
  };

  const [row] = await db
    .insert(weeklyReviews)
    .values(values)
    .onConflictDoUpdate({
      target: weeklyReviews.weekStart,
      set: {
        createdAt,
        dsa: values.dsa,
        lld: values.lld,
        hld: values.hld,
        ai: values.ai,
        personal: values.personal,
      },
    })
      .returning();

  if (!row) throw new Error("Failed to save weekly review");
  return serializeReview(row);
}

export async function recentSessions(limit = 20): Promise<SessionRecord[]> {
  const rows = await db
    .select()
    .from(sessions)
    .orderBy(desc(sessions.ts), desc(sessions.id))
    .limit(limit);
  return rows.map(serializeSession);
}

export async function getReview(weekStart: string): Promise<ReviewRecord | null> {
  const [row] = await db
    .select()
    .from(weeklyReviews)
    .where(eq(weeklyReviews.weekStart, weekStart))
    .limit(1);
  return row ? serializeReview(row) : null;
}

export async function statsForWeek(moment: Date): Promise<WeekStats> {
  const timeZone = loadConfig().timezone;
  const start = weekStartIso(moment, timeZone);
  const [sy, sm, sd] = start.split("-").map(Number);
  const end = addCalendarDays(sy, sm, sd, 7);
  const startDt = zonedWallToDate(sy, sm, sd, 0, 0, timeZone);
  const endDt = zonedWallToDate(end.year, end.month, end.day, 0, 0, timeZone);

  const [rows, problems, review] = await Promise.all([
    db
      .select()
      .from(sessions)
      .where(and(gte(sessions.ts, startDt), lt(sessions.ts, endDt))),
    listProblems(),
    getReview(start),
  ]);

  const bySubject: Record<string, SubjectBucket> = {};
  const walkDays = new Set<string>();
  const readingDays = new Set<string>();

  for (const row of rows) {
    const bucket = (bySubject[row.subject] ??= {
      minutes: 0,
      sessions: 0,
      problems: 0,
    });
    bucket.minutes += row.minutes || 0;
    bucket.sessions += 1;
    bucket.problems += row.problemsCount || 0;
    const day = ymdInZone(row.ts, timeZone);
    if (row.subject === "walk") walkDays.add(day);
    if (row.subject === "reading") readingDays.add(day);
  }

  const solved = weekDsaProblemStats(
    problems.map((problem) => problem.lastSolved),
    start,
  );

  return {
    week_start: start,
    by_subject: bySubject,
    walk_days: walkDays.size,
    reading_days: readingDays.size,
    dsa_problems_total: solved.dsa_problems_total,
    dsa_problems_week: solved.dsa_problems_week,
    study_minutes_week: ["dsa", "lld", "hld", "ai"].reduce(
      (sum, key) => sum + (bySubject[key]?.minutes ?? 0),
      0,
    ),
    review,
  };
}
